crates/wasi/src/snapshots/preview_1.rs

```diff
@@ -0,0 +1,513 @@
+
+// Creates a dummy `RawWaker`. We can only create Wakers from `RawWaker`s
+fn dummy_raw_waker() -> RawWaker {
+    fn no_op(_: *const ()) {}
+    //returns a new RawWaker by calling dummy_raw_waker again
+    fn clone(_: *const ()) -> RawWaker {
+        dummy_raw_waker()
+    }
+    // RawWakerVTable specifies the functions that should be called when the RawWaker is cloned, woken, or dropped.
+    let vtable = &RawWakerVTable::new(clone, no_op, no_op, no_op);
+
+    RawWaker::new(std::ptr::null::<()>(), vtable)
+}
+
+// Creates a dummy waker which does *nothing*, as the future itsef polls to ready at first poll
+// A waker is needed to do any polling at all, as it is the primary constituent of the `Context` for polling
+fn run_in_dummy_executor<F: std::future::Future>(f: F) -> Result<F::Output, Trap> {
+    let mut f = Pin::from(Box::new(f));
+    let waker = unsafe { Waker::from_raw(dummy_raw_waker()) };
+    let mut cx = Context::from_waker(&waker);
+    match f.as_mut().poll(&mut cx) {
+        std::task::Poll::Ready(val) => Ok(val),
+        std::task::Poll::Pending => Err(Trap::new("Cannot wait on pending future")),
+    }
+}
+
+// Creates the function item `add_wasi_snapshot_preview1_to_wasmi_linker` which when called adds all
+// `wasi preview_1` functions to the linker
+macro_rules! impl_add_to_linker_for_funcs {
+    (
+        $(
+            $( #[$docs:meta] )*
+            fn $fname:ident ($( $arg:ident : $typ:ty ),* $(,)? ) -> $ret:tt
+        );+ $(;)?
+    ) => {
+        fn add_wasi_snapshot_to_wasmi_linker<'a, T, U>(
+            linker: &mut Linker<T>,
+            mut store_ctx: impl AsContextMut<UserState = T>,
+            wasi_ctx: impl Fn(&mut T) -> &mut U + Send + Sync + Copy + 'static)
+        -> Result<(), Error>
+        where U: wasi_common::snapshots::preview_1::wasi_snapshot_preview1::WasiSnapshotPreview1 +
+                 wasi_common::snapshots::preview_1::wasi_snapshot_preview1::UserErrorConversion
+        {
+            let mut store = store_ctx.as_context_mut();
+            $(
+                $(#[$docs])*
+                linker.define(
+                    "wasi_snapshot_preview1",
+                    stringify!($fname),
+                    Func::wrap(&mut store, move|mut caller: Caller<'_, T>, $($arg : $typ,)*| {
+                        let result = async {
+                            let mem = match  caller.get_export("memory") {
+                                Some(Extern::Memory(m)) => m,
+                                _ => return Err(Trap::new("missing required memory export".to_string())),
+                            };
+                            let(mem, ctx) = mem.data_and_store_mut(&mut caller);
+                            let ctx = wasi_ctx(ctx);
+                            let mem = wiggle::wasmtime::WasmtimeGuestMemory::new(mem);
+
+                            match wasi_common::snapshots::preview_1::wasi_snapshot_preview1::$fname(ctx, &mem, $($arg,)*).await {
```

Comment:
I think you cannot use `await` in a non-`async` function. Instead you have to use blocking semantics to block on the `async` function in `wasi_common`. I wonder why this is not reported by the compiler atm. Maybe because we don't test it yet?

Reply:
Hmmn. The function being called itself is an async function, thus should be awaited, even if blocked on. 

What you said is true. I'll have to look into this.

Reply:
Oh. Look at me. Was replying from my phone, so  didn't check the code. The await call is inside an `async {...}` block. The `future` which the block evaluates to is then run inside our dummy executor which simply blocks on it. So, we're fine.